analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs

```diff
@@ -168,133 +144,275 @@ public static void LoopWithBreak(System.Collections.Generic.IEnumerable<string>
+        foreach ((AStruct?, AStruct) x in
+            new (AStruct?, AStruct?)[] { (new AStruct(), new AStruct()), (new AStruct(), null) }) ; // FN, value access on second item of the couple
+    }
+
+    void ForeachDestructuring()
+    {
+        foreach ((var x, var y) in new (AStruct?, AStruct?)[] { (new AStruct(), null) }) ;
+        foreach ((AStruct? x, AStruct y) in new (AStruct?, AStruct?)[] { (new AStruct(), null) }) ; // Error[CS0266]
+        foreach ((var x, AStruct y) in new (AStruct?, AStruct?)[] { (new AStruct(), null) }) ;      // Error[CS0266]
     }
+
+    private struct AStruct { }
 }
 
-// https://github.com/SonarSource/sonar-dotnet/issues/4573
-public class Repro_4573
+class Arithmetic
 {
-private DateTime? foo;
+    int? MultiplicationByZeroAndAssignment(int? i) => 0 * (i = null) * i.Value;    // Noncompliant
+    int? MultiplicationAndAssignments1(int? i) => (i = 42) * (i = null) * i.Value; // Noncompliant
+    int? MultiplicationAndAssignments2(int? i) => (i = null) * (i = 42) * i.Value; // Compliant
+}
 
-public virtual DateTime? Foo
+class ComplexConditionsSingleNullable
 {
-    get => foo;
-    set
-    {
-        if (value.HasValue)
-        {
-            //HasValue and NoValue constraints are set here
-        }
-        if (foo != value || foo.HasValue)
-        {
-            foo = value;
-        }
-    }
+    bool LogicalAndLearningNonNull1(bool? b) => b.HasValue && b.Value;
+    bool LogicalAndLearningNonNull2(bool? b) => b.HasValue == true && b.Value;
+    bool LogicalAndLearningNonNull3(bool? b) => b.HasValue != false && b.Value;
+    bool LogicalAndLearningNonNull4(bool? b) => b.HasValue == !false && b.Value;
+    bool LogicalAndLearningNonNull5(bool? b) => !!b.HasValue && b.Value;
+
+    bool LogicalAndLeaningNull1(bool? b) => !b.HasValue && b.Value;         // Noncompliant - parameter should be null-constrained
+    bool LogicalAndLeaningNull2(bool? b) => b.HasValue == false && b.Value; // Noncompliant
+    bool LogicalAndLeaningNull3(bool? b) => b.HasValue == !true && b.Value; // Noncompliant
+    bool LogicalAndLeaningNull4(bool? b) => !!!b.HasValue && b.Value;       // Noncompliant
+
+    bool BitAndLearningNull(bool? b) => !b.HasValue & b.Value;              // Noncompliant - i.Value reached in any case
+    bool BitAndLearningNonNull(bool? b) => b.HasValue & b.Value;            // Noncompliant - i.Value reached in any case
+
+    bool LogicalOrLearningNonNull(bool? b) => b.HasValue || b.Value;        // Noncompliant - i.Value reached when i.HasValue is false
+
+    bool BitOrLearningNonNull(bool? b) => b.HasValue | b.Value;             // Noncompliant - i.Value reached in any case
+
+    bool Tautology(bool? b) => (!b.HasValue || b.HasValue) && b.Value;      // Noncompliant
+
+    bool ShortCircuitedOr(bool? b) => (true || b.HasValue) && b.Value;      // Compliant
+    bool ShortCircuitedAnd(bool? b) => (false &&  b.HasValue) || b.Value;   // Compliant
```

Comment:
These might be 'FN` that we don't want to fix. As user indicated it could be null, and accesses it in any case.

S2583 should kick in to warn users about the problem itself.